streaming/sender.c

```diff
@@ -714,3 +733,128 @@ void *rrdpush_sender_thread(void *ptr) {
+{
+    UNUSED(s);
+    RRDDIM *rd;
+    struct rrddim_query_handle handle;
+
+    // When the sync is unknown against the far end (last_sent=0), sending the latest sample will trigger a
+    // replication from the far end if necessary.
+    time_t first_t = rrdset_first_entry_t(st);
+    time_t st_newest = st->last_updated.tv_sec;
+    time_t window_start;
+
+    if (start_time == 0)
+        window_start = MAX((time_t)st->state->last_sent.tv_sec + st->update_every, first_t);
+    else
+        window_start = MAX(start_time, first_t);
+
+    // Decide how much data to replicate from the beginning of the window
+    time_t unsent_points = (st_newest - window_start) / st->update_every + 1;
+    if (unsent_points > default_rrdpush_gap_block_size)
+        unsent_points = default_rrdpush_gap_block_size;
+    time_t window_end = window_start + unsent_points * st->update_every;    // window_end <= st_newest + update_every
+
+    // If we are responding to an explicit request then we may send empty windows, these communicate to the far end
+    // that there is no data in the interval.
+    if (start_time == 0)
+        buffer_sprintf(s->build, "REPBEGIN %s %ld %ld\n", st->id, window_start, window_end);
+    else
+        buffer_sprintf(s->build, "REPBEGIN %s %ld %ld\n", st->id, start_time, window_end);
+
+    size_t num_points = 0;
+    rrddim_foreach_read(rd, st) {
+        // Send the intersection of this dimension and the time-window on the chart
+        if (!rd->exposed)
+            continue;
+        time_t rd_start = rrddim_first_entry_t(rd);
+        time_t rd_end   = rrddim_last_entry_t(rd) + st->update_every;
+        if (rd_start < window_end && rd_end >= window_start) {
+
+            time_t rd_oldest = MAX(rd_start, window_start);
+            rd_end = MIN(rd_end,   window_end);
+
+            time_t sample_t, metric_t;
+            rd->state->query_ops.init(rd, &handle, rd_oldest, rd_end);
+            debug(D_REPLICATION, "Fill replication with %s.%s @%ld-%ld rd@%ld-%ld", st->id, rd->id, window_start,
+                                 window_end, rd_oldest, rd_end);
+
+            // Technically rd->update_every could differ from st->update_every, but it does not.
+            for( sample_t = rd_oldest; sample_t < rd_end; sample_t += rd->update_every) {
+
+                if (rd->state->query_ops.is_finished(&handle)) {
+                    debug(D_REPLICATION, "%s.%s query handle finished early @%ld", st->id, rd->id, sample_t);
+                    break;
+                }
+
+                storage_number n = rd->state->query_ops.next_metric(&handle, &metric_t);
+                if (n==SN_EMPTY_SLOT) {
+                    debug(D_REPLICATION, "%s.%s db empty in valid dimension range @ %ld", st->id, rd->id, sample_t);
+                }
+                if (sample_t != metric_t) {
+                    debug(D_REPLICATION, "%s.%s Sample mismatch during replication %ld vs %ld", st->id, rd->id,
```

Comment:
Data replication cannot depend on `update_every`. `dbengine` supports variable granularity in data collection. As a result, a query can return results that contain gaps and various frequencies, e.g.:

| | Data points |
| ------------- | ------------- |
| frequency |  1, 1, 1, 1, GAP, ... GAP, 2, 2, 2, 2, 2, 2, 2, GAP, ... GAP, 3 |

You should remove the `for` loop condition `sample_t < rd_end` since you can potentially lose data points this way. `sample_t` does not seem like it's giving useful information in general.

Reply:
Moreover, it is possible that the `next_metric()` call will jump in time, so the frequency change must be inferred.

Reply:
Fixed